activerecord/lib/active_record/associations/collection_proxy.rb

```diff
@@ -1110,6 +1111,7 @@ def reload
       def reset
         proxy_association.reset
         proxy_association.reset_scope
+        @offsets = {}
         reset_scope
```

Comment:
I initially added it to `reset_scope`, but since `reset_scope` is called [each time the collection proxy is accessed](https://github.com/rails/rails/blob/ab0e4558816b1a69b4e96446ffe2812e21053b42/activerecord/lib/active_record/associations/collection_association.rb#L34), we end up not caching the offsets on the association at all. The two specs I added catch this with their first assertion.